### PR TITLE
Add foreign key join. Issue #364

### PR DESCRIPTION
**Pull request by raymcdermott**
Add support for the `join` arity of a `ktable` that accepts a `foreign-key-extractor` function

## Blog post that introduces the feature

https://www.confluent.io/blog/data-enrichment-with-kafka-streams-foreign-key-joins/

## Official API

https://kafka.apache.org/24/javadoc/org/apache/kafka/streams/kstream/KTable.html#join-org.apache.kafka.streams.kstream.KTable-java.util.function.Function-org.apache.kafka.streams.kstream.ValueJoiner-

# Checklist

- [X] tests
- [X] updated [CHANGELOG](../CHANGELOG.md) (the "unreleased" section)

### CODE REVIEW

**Comment by fc-infosec-sast** [comment]
<h2>✅ SAST Report</h2>

No issues found. Good job! 💪

File: CHANGELOG.md [hunk 1]
```diff
@@ -2,6 +2,9 @@
 
 ### Unreleased
 
+### [0.9.12] - [2023-11-24]
+- Support for Foreign Key joins (Issue #364)
```
Comment (minimal) [inline CHANGELOG.md:6]: this should be in unreleased

File: test/jackdaw/streams_test.clj [hunk 2]
```diff
@@ -725,8 +749,8 @@
 
           ;; if we don't set the `grace` period of the `TimeWindows`, the
           ;; default is used: 24h - window
-          window       (Duration/ofMillis 100)
-          grace        (Duration/ofMillis 1)
+          window (Duration/ofMillis 100)
```
Comment (minimal) [inline test/jackdaw/streams_test.clj:752]: Would prefer not to realign unrelated code.

**Comment by minimal** [comment]
LTGM aside from the minor comments

File: test/jackdaw/streams_test.clj (hunk 2 above)
Comment (raymcdermott) [reply test/jackdaw/streams_test.clj:752]: 🤦🏻

Comment (raymcdermott) [reply test/jackdaw/streams_test.clj]: Fixed

**Review by minimal — APPROVED** [review]
Thanks!
